MAINT: make special.factorial handle nan correctly

rlucas7 opened the pull request:
#### Reference issue
Closes #11022 

#### What does this implement/fix?
Previously, `scipy.special.factorial` did not always handle `nan` correctly. 
This PR fixes the `nan` behavior to propogate as expected.

#### Additional information
There are added unit tests too that account for `nan` scalar, `np.array([np.nan])` and mixed arrays, e.g. `np.array([np.nan, 1, 2., np.nan])`. All 3 cases are handled as expected.

rlucas7 commented:
Hmm, all tests passed for me locally... looks like the `un[un > 1]` throws an error when `greater` comparator is used on `np.nan` (my guess) I'll put the line with this into the `errstate` context manager too and see if that resolves the build issue.

rlucas7 commented:
The travis fail is due to a single build, it looks like one of the workers failed during the build.

Direct link: 
https://travis-ci.org/scipy/scipy/jobs/627551280#L1600
everything else is green.

person142 commented on scipy/special/_basic.py:
```diff
@@ -2036,13 +2036,18 @@ def factorial(n, exact=False):
+            if not np.isnan(n):
+                return 0 if n < 0 else math.factorial(n)
+            else:
+                return np.nan
```
In this case return `n`. Reason: there are actually many NaNs and you want to propagate them if possible.

rlucas7 replied:
ok-good point-I hadn't considered multiple NaNs scenarios.

pvanmulbregt replied:
The `else:` is unneeded as both branches immediately return.  E.g. (also incorporating Josh's comment)
```
if np.ndim(n) == 0:
  if np.isnan(n):
    return n
 return 0 if n < 0 else math.factorial(n)
```

rlucas7 replied:
@pvanmulbregt thanks for the feedback. I think with the `_factorial(n)` ufunc we no longer need the block. In other words: 
```
if np.ndim(n) == 0:
  if np.isnan(n):
    return n
 return 0 if n < 0 else math.factorial(n)
```
becomes:
```
if np.ndim(n) == 0:
    return _factorial(n)
```
Further simplifications in the other parts of the code were possible. 
I removed quite a bit of the top level `else:` block `nan` handling.

person142 commented on scipy/special/_basic.py:
```diff
@@ -2065,14 +2072,27 @@ def factorial(n, exact=False):
     else:
         if np.ndim(n) == 0:
-            return 0 if n < 0 else gamma(n + 1)
+            if not np.isnan(n):
```
The check shouldn't be needed here-`gamma` will propagate the NaN correctly.

rlucas7 replied:
yes, I got overzealous in my NaN handling.

person142 commented on scipy/special/_basic.py:
```diff
@@ -2065,14 +2072,27 @@ def factorial(n, exact=False):
         n = asarray(n)
         vals = gamma(n + 1)
-        return where(n >= 0, vals, 0)
+        # Ignore runtime warning when less operator used w/np.nan
```
We can simplify here by reworking the logic a bit-do something like

```python
out = np.empty_like(n)
with np.errstate(all='ignore'):
    neg = n < 0
out[neg] = 0
out[~neg] = gamma(n + 1)
```

Since `gamma` will propagate NaNs correctly we don't have to worry about that part.

Even better would be to make a tiny private ufunc with a kernel like:

```cython
cdef double factorial(double n) nogil:
    if n < 0:
        return 0
    else:
        return gamma(n + 1)
```

and then this entire `else` branch can just be calling that function.

rlucas7 replied:
@person142
I like the private ufuncs idea plus that may be a useful pattern to follow in future maintenance work in special. 

Should the ufunc go under cephes or somewhere else? (if somewhere else I'm not sure where)

person142 replied:
I would default to Cython (we have a vague hope that Cython is more maintainable for people that don't know C) and currently we just put all Cython stuff in the `special` directory, something like `special/_factorial.pxd`. You can get at `gamma` by doing a

```cython
from ._cephes cimport gamma
```

I imagine that ultimately most of the factorial implementation would benefit from being rewritten in Cython, so this would be a good start for that too.

rlucas7 replied:
@person142 I started doing this and ran into some build failures, not sure if i'm missing something. I was cribbing off this PR: 
https://github.com/scipy/scipy/pull/10440/files
does that look like what you had in mind (modulo the changes to be fore factorial)? 

I though I was, but maybeI'm still not clear on the ***full*** build process for `scipy.special`.

person142 replied:
Yeah, that PR should be fairly representative. You can always push a partial commit if further debugging is needed.

rlucas7 replied:
I figured it out, I was missing the empty, "internal do not use" docs which cause the build to fail. After that and a couple other minor fails I was able to rebuild and do some fixups/cleanups w/`_factorial`.

rlucas7 commented:
In addition to changes to address the PR comments I also rebased off master in the hope that recent changes Ralf and Tyler put in to fix the failing builds help this PR get to green. Let's see what the CI tools say.

rlucas7 commented:
Looks like one of the doctests (maybe?) picked up a type change with these changes:
https://travis-ci.org/scipy/scipy/jobs/632545920#L1243

expected output `120L` got `120.0`. I'll change this, tomorrow.

person142 commented on scipy/special/_basic.py:
```diff
@@ -14,7 +14,7 @@
 from ._ufuncs import (mathieu_a, mathieu_b, iv, jv, gamma,
                       psi, hankel1, hankel2, yv, kv, ndtri,
-                      poch, binom, hyp0f1)
+                      poch, binom, hyp0f1, _factorial)
```
p.s. just use the above `ufuncs` import and do `ufuncs._factorial`. Generally trying to cut down on the number of un-namespaced functions in `_basic.py`.

rlucas7 replied:
ah yeah I see your point. I hadn't considered it would be in there now with the import-changed.

person142 commented on scipy/special/_basic.py:
```diff
@@ -2036,13 +2036,15 @@ def factorial(n, exact=False):
     if exact:
         if np.ndim(n) == 0:
-            return 0 if n < 0 else math.factorial(n)
+            return _factorial(n)
```
Since we're in the exact branch you'll still want to use `math.factorial` here so that it handles very large integers.

person142 commented on scipy/special/_factorial.pxd:
```diff
@@ -0,0 +1,8 @@
+from ._cephes cimport Gamma
+
+
+cdef inline double _factorial(double n) nogil:
```
So here's a question: should this signature be `double _factorial(long double n)`? You could definitely get doubles through before, but I'm tempted to call that behavior broken. I'm all for pushing people to use `sc.gamma` if they actually want results for non-integral values.

rlucas7 replied:
Hmm, I don't think that will lead to consistent behavior in `_factorial(n)` and then we'll need more type checking in `exact=True` branch, where `math.factorial(5.0)` will return `120` without error or warning.  

```
>>> import math
>>> type(math.factorial(5.0))
<class 'int'>
```
I'm not opposed to forcing input `n` to be `int` in `_factorial`, but if we do that I think it should be consistent behavior within the function. For example, currently, `math.factorial(5.1)` throws a `ValueError` but `math.factorial(5.0)` is the correct `int` returned. 
```
>>> math.factorial(5.1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: factorial() only accepts integral values
```
We'd either want to do something like in the legacy deprecations (e.g. the truncation in `bdtr` etc.) but here would be ***without*** a `RuntimeWarning` emitted. Or we could remove the `math.factorial(n)` call to get consistent behavior across the 2 branches of `exact`. Unless you see a third way that I'm not thinking about at the moment. 

Does that make sense? 

If so, I'm pretty much indifferent to the approach to get us there, so long as the two branches are consistent. 

If we're on the same page, which approach would you recommend?

rlucas7 replied:
I looked at this today and is more clear how to proceed. So I'll do a check similar to:
https://github.com/scipy/scipy/blob/master/scipy/special/_legacy.pxd#L33-L37
but in a separate function and return a `ValueError` if `<int> x != x` I'll call the function something like `_int_arg_check()` and have the return message match that of the `math.factorial` this way we have consistent messaging on value errors across the `exact=True` and `exact=False` branches.

rlucas7 replied:
Actually, what seems simpler is to put it directly into the `_factorial()` in cython, but it seems that the `ValueError` is ignored when thrown? I'm guessing this is some property of the build setup but I'm not clear if there is a way to override the behavior specifically for `_factorial()` or if it is better to place up in the python itself. @person142 can you give me a pointer on this? 

I pushed my changes if that helps to look at to understand. The message I get in the test is: 
```
===================================================================================== FAILURES ======================================================================================
__________________________________________________________________ TestFactorialFunctions.test_non_integral_inputs __________________________________________________________________
scipy/special/tests/test_basic.py:1842: in test_non_integral_inputs
    assert_raises(ValueError, special.factorial, 1.1, False)
E   Failed: DID NOT RAISE <class 'ValueError'>
------------------------------------------------------------------------------- Captured stderr call --------------------------------------------------------------------------------
ValueError: factorial() only accepts integral values
Exception ignored in: 'scipy.special._factorial._factorial'
ValueError: factorial() only accepts integral values
```

person142 replied:
Ok, so sorry to waffle here but yeah, I think your first instinct of "let's not change this" is correct, especially for this PR. The SciPy factorial function has been weird for a long time, so we shouldn't try and cram a fix for that in here.

rlucas7 replied:
Ok, so to be clear, I'll revert the last commit and address the couple (minor) comments you have.

rlucas7 commented:
Did a small rebase and reverted the `exact` branch to use `math.factorial` and removed the direct import. Still need to align with @person142 on approach for signature (and return type) of `_factorial` in Cython.

person142 commented on scipy/special/_basic.py:
```diff
@@ -2353,14 +2359,15 @@ def factorial(n, exact=False):
-        if np.ndim(n) == 0:
-            return 0 if n < 0 else gamma(n + 1)
-
         n = asarray(n)
```
Now you can delete the `assaray` line. Really this branch can just be `return ufuncs._factorial(n)`.

rlucas7 commented:
@person142 hopefully I understood your comments here correctly. I reverted the last commit w/signature mucking and also removed the `asarray(n)` based on your comment. I was a bit behind master so I also did another rebase. I did not squash commits as I assume it is easier to see the commits preserved. This passed all tests locally so lets see what CI shows.

person142 commented on scipy/special/_basic.py:
```diff
@@ -2352,14 +2358,14 @@ def factorial(n, exact=False):
                     val *= _range_prod(prev, current)
                     out[n == current] = val
+
+            if np.isnan(n).any():
```
Maybe in a follow up store the `np.isnan` array so that it doesn't get computed so many times.

person142 commented:
Merged, thanks @rlucas7.
